File: lib/src/main/java/one/hyro/builders/CustomItem.java
package one.hyro.builders;

import lombok.Getter;
import net.kyori.adventure.text.Component;
import one.hyro.managers.MenuManager;
import one.hyro.utils.CustomHeads;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;

import java.util.List;
import java.util.function.Consumer;

@Getter
public class CustomItem {
    private String customId;
    private Material material;
    private ItemStack item;
    private ItemMeta meta;
    private Consumer<Player> consumer;

    public CustomItem(Material material) {
        this.material = material;
        this.item = new ItemStack(material);
        this.meta = item.getItemMeta();
    }

    public CustomItem(String texture) {
        this.item = CustomHeads.getCustomHead(texture);
        this.meta = (SkullMeta) item.getItemMeta();
    }

    public CustomItem(Player player) {
        this.item = CustomHeads.getPlayerHead(player);
        this.meta = (SkullMeta) item.getItemMeta();
    }

    public CustomItem setCustomId(String customId) {
        this.customId = customId;
        return this;
    }

    public CustomItem setDisplayName(Component component) {
        meta.displayName(component);
        return this;
    }

    public CustomItem setLore(Component components) {
        meta.lore(List.of(components));
        return this;
    }

    public CustomItem amount(int amount) {
        if (amount < 1) amount = 1;
        item.setAmount(amount);
        return this;
    }

    public CustomItem setEnchantment(Enchantment enchantment, int level) {
        meta.addEnchant(enchantment, level, false);
        return this;
    }

    public CustomItem hideEnchantments() {
        meta.addItemFlags(ItemFlag.HIDE_ENCHANTS);
        return this;
    }

    public CustomItem onClick(Consumer<Player> consumer) {
        this.consumer = consumer;
        return this;
    }

    public CustomItem build() {
        item.setItemMeta(meta);
        MenuManager.getInstance().registerItem(this);
        return this;
    }
}

File: lib/src/main/kotlin/builder/inventory/CustomItem.kt
package one.hyro.builder.inventory

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextDecoration
import one.hyro.Lib
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.persistence.PersistentDataType

class CustomItem(material: Material) {
    var item: ItemStack = ItemStack(material)
    var meta: ItemMeta? = item.itemMeta
    var interactionConsumer: ((Player) -> Unit)? = null
    var clickConsumer: ((Player) -> Unit)? = null

    fun amount(amount: Int) = apply {
        if (amount <= 0 || amount > 64) item.amount = 1 else item.amount = amount
        return this
    }

    fun displayName(displayName: Component) = apply {
        meta?.displayName(displayName.decoration(TextDecoration.ITALIC, false))
        item.itemMeta = meta
        return this
    }

    fun lore(lore: Component) = apply {
        meta?.lore(listOf(lore.decoration(TextDecoration.ITALIC, false)))
        item.itemMeta = meta
        return this
    }

    fun enchantment(enchantment: Enchantment, level: Int) = apply {
        item.addEnchantment(enchantment, level)
        return this
    }

    fun hideEnchantments() = apply {
        item.itemMeta?.addItemFlags(ItemFlag.HIDE_ENCHANTS)
        return this
    }

    fun interact(consumer: (Player) -> Unit) = apply {
        this.interactionConsumer = consumer
        return this
    }

    fun click(consumer: (Player) -> Unit) = apply {
        this.clickConsumer = consumer
        return this
    }

    fun persistentData(key: String, value: String) = apply {
        val namespacedKey = NamespacedKey(Lib.plugin, key)
        meta?.persistentDataContainer?.set(namespacedKey, PersistentDataType.STRING, value)
        return this
    }

    fun build() = apply {
        item.itemMeta = meta
        return this
    }
}